docs/extensions/guides/working-with-mobx.md

```diff
@@ -0,0 +1,78 @@
+# Working with mobx
+
+## Introduction
+
+Lens uses `mobx` as its state manager on top of React's state management system.
+This helps with having a more declarative style of managing state, as opposed to `React`'s native `setState` mechanism.
+You should already have a basic understanding of how `React` handles state ([read here](https://reactjs.org/docs/faq-state.html) for more information).
+However, if you do not, here is a quick overview.
+
+A `React.Component` is generic over both `Props` and `State` (with default empty object types).
+`Props` should be considered read-only from the point of view of the component and is the mechanism for passing in "arguments" to a component.
+`State` is a component's internal state and can be read by accessing the parent field `state`.
+`State` **must** be updated using the `setState` parent method which merges the new data with the old state.
+`React` does do some optimizations around re-rendering components after quick successions of `setState` calls.
+
+## How mobx works:
+
+`mobx` is a package that provides an abstraction over `React`'s state management. The three main concepts are:
+- `observable`: data stored in the component's `state`
+- `action`: a function that modifies any `observable` data
+- `computed`: data that is derived from `observable` data but is not actually stored. Think of this as computing `isEmpty` vs an `observable` field called `count`.
+
+Further reading is available from `mobx`'s [website](https://mobx.js.org/the-gist-of-mobx.html).
+
+## Basic usage of mobx:
+
+When using `Lens`'s extension's API, some of the provided types are marked as `observable` (or others) and are documented as such.
+These can be used as normal types and the combination of `mobx` and `react` work to determine when a rerender should occur.
+
+---
+
+## Example:
+
+Imagine that your extension wants to conditionally add an app menu if some other observable value is ever `true` (and remove it when it is `false`).
+That could be achieved using roughly the following code:
+
+```typescript
+import { LensMainExtension, MenuRegistration } from "@k8slens/extensions";
+import observables from "./observables" // a collection of observable data
+
+interface MenuRegistrationWithId extends MenuRegistration {
+  id?: string;
+}
+
+export default class ExtensionMain extends LensMainExtension {
+  appMenus: MenuRegistrationWithId[] = [
+    {
+      parentId: "file",
+      label: strings.extension.appMenu["label"](),
+      click() {
+        console.log("you clicked the label menu")
+      },
+    },
+  ];
+
+  constructor() {
+    super()
+
+    reaction(
+      () => observables.clusterIsInState,
+      clusterIsInState => {
+        if (clusterIsInState) {
+          this.appMenus.push({
```

Comment:
What happens here? Does this trigger ExtensionLoader.init() by virtue of appMenus being an observable?

Reply:
That is what I gathered from looking around. Yes, that is what the observable implies.

Reply:
1. `ExtensionLoader.init` is called only once when app is started

1. `this.appMenus.push({})` probably won't trigger anything since observable only a link to array itself?
```ts
    this.autoInitExtensions((ext: LensMainExtension) => [
      registries.menuRegistry.add(ext.appMenus)
    ]);
```
Otherwise if extension is enabled and already added some menus it should de-register items first before modification?

Reply:
So this example doesn't work anyway?

Reply:
Probably.. (haven't time to check)

Reply:
So maybe it is disingenuous to mark `appMenus` as `observable` then. Since we never actually "observe" the actions.

Reply:
Or we should add such support and make them truly observable? :)

Reply:
I think we should, and it was asked for in #1375. If we do want to work on it, then this PR shouldn't be merged until after that work is done (or it doesn't matter if both are finished by 4.0.0)